### PR TITLE
[FEATURE] HardSubX Russian Tickers

### PR DESCRIPTION
**Pull request by BurhanUlTayyab**
Please prefix your pull request with one of the following: **[FEATURE]** **[FIX]** **[IMPROVEMENT]**.

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I have considered, and confirmed that this submission will be valuable to others.
- [x] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [x] I give this submission freely, and claim no ownership to its content.

**My familiarity with the project is as follows (check one):**

- [ ] I have never used CCExtractor.
- [ ] I have used CCExtractor just a couple of times.
- [x] I absolutely love CCExtractor, but have not contributed previously.
- [ ] I am an active contributor to CCExtractor.

---

Merging my GSoC'18 work on Extracting Russian News Tickers and combining them in correct order (Early and Late Fusion Algorithm). This work was done as a partnership between Red Hen Labs and CCExtractor (as I'm from Red Hen Labs and my mentor is Abhinav Shukla).

### CODE REVIEW

File: src/lib_ccx/hardsubx_decoder.c [hunk 1]
```diff
@@ -11,6 +11,19 @@
 #include "hardsubx.h"
 #include "capi.h"
 
+
+
+
+
+
+
```
Comment (MatejMecka) [inline src/lib_ccx/hardsubx_decoder.c]: Why a bunch of new lines?

File: src/lib_ccx/hardsubx.h [hunk 2]
```diff
@@ -69,6 +69,18 @@ struct lib_hardsubx_ctx
 
 	// Subtitle text parameters
 	int tickertext;
+        int russian;
+        int frame_skip;
```
Comment (MatejMecka) [inline src/lib_ccx/hardsubx.h]: make the indentation the same level? It looks weird now

File: src/lib_ccx/hardsubx.c [hunk 3]
```diff
@@ -255,6 +260,17 @@ struct lib_hardsubx_ctx* _init_hardsubx(struct ccx_s_options *options)
 	ctx->write_format = options->write_format;
 	ctx->subs_delay = options->subs_delay;
 	ctx->cc_to_stdout = options->cc_to_stdout;
+        ctx->russian = options->russian;
+        ctx->frame_skip = options->frame_skip;
+        ctx->start_frame = options->start_frame;
```
Comment (MatejMecka) [inline src/lib_ccx/hardsubx.c]: make the indentation the same level? It looks weird now

**Comment by cfsmp3** [comment]
I see *lots* of magic numbers and hardcoded stuff, why? Seems like it's only going to work for some specific use cases...

**Comment by BurhanUlTayyab** [comment]
I definitely used global variables but they are default. Other than that, I have created a specific file titled "hardsubx_russian.c" which contains Early and Late Fusion Algorithm for combining tickers in correct order. There's only thing that is currently hardcoded and that is the detection part and on that Saurabh is currently working on and I'm also talked about that to Abhinav and Francis Steen from Red Hen Labs. Other than that, everything is variable and depends upon input

**Comment by cfsmp3** [comment]
Everything there seems magic to me:
	if (X > 0.008856)
		fX = pow(X, 1.0/3.0);
	else
		fX = 7.787*X + 16.0/116.0;

Where does each of those numbers come from?

**Comment by BurhanUlTayyab** [comment]
Also I have also modified and reused a lot of Abhinav's HardSubX Code to fit Russian Text Extraction from News. Kindly tell me the stuff that is hardcoded other than Box (detection part)

**Comment by BurhanUlTayyab** [comment]
Regarding that, I believe that is Abhinav's work, and that is already part of CCExtractor for a very long time. Link here (https://github.com/CCExtractor/ccextractor/blob/master/src/lib_ccx/hardsubx_imgops.c). I just modified the function to new threshold value and renamed it, the majority of the work is Abhinav's Work when he was GSoC Student

**Comment by Abhinav95** [comment]
@cfsmp3 The magic numbers in the colorspace function are industry standard specs for RBG to LAB conversion, nothing wrong there. There are a few other hardcoded parameters specific to the tickers in the Russian news channels that we want to support, which will hopefully be solved by @saurabhshah0410's PR which should be coming soon.

I do see some other places where this PR needs improvement though, which I will mention in the specific review places.

@BurhanUlTayyab Needs a little more work before merging.

File: src/lib_ccx/hardsubx_decoder.c [hunk 4]
```diff
@@ -33,8 +34,8 @@ char* _process_frame_white_basic(struct lib_hardsubx_ctx *ctx, AVFrame *frame, i
 			int b=frame->data[0][p+2];
 			pixSetRGBPixel(im,j,i,r,g,b);
 			float L,A,B;
-			rgb_to_lab((float)r,(float)g,(float)b,&L,&A,&B);
-			if(L > ctx->lum_thresh)
+			rgb2lab((float)r,(float)g,(float)b,&L,&A,&B);
+			if(L > 85)
```
Comment (Abhinav95) [inline src/lib_ccx/hardsubx_decoder.c]: This is an incorrect and unnecessary change. Please also do not change the naming convention of rgb_to_lab

File: src/lib_ccx/hardsubx_russian.c [hunk 5]
```diff
@@ -0,0 +1,995 @@
...
+         last_index = index;
+         //subtitle_text = russian_ocr(pixd, index);
+         printf("%f", check_frames(pixd, pix_temp));
+
+         if (check_frames(pixd, pix_temp) <= 30){
+         t_flag = 0;
+         sprintf(write_path,"im%04d.jpg",pre_index);
+         if (count_frame % 2 == 0) {
+         pixWrite(write_path,pix_temp,IFF_JFIF_JPEG);
+         subtitle_text = russian_ocr(pixd, index);
```
Comment (Abhinav95) [inline src/lib_ccx/hardsubx_russian.c]: Please write to file only in debug mode (within #ifdef DEBUG). Also, do not use printf anywhere for non useful information ("Here"). To print useful information, use mprint() as elsewhere in the code.

File: src/lib_ccx/hardsubx_russian.c [hunk 6]
```diff
@@ -0,0 +1,995 @@
...
+	PIX *im;
+	PIX *edge_im;
+	PIX *lum_im;
+	PIX *feat_im;
+	char *subtitle_text=NULL;
+	im = pixCreate(width,height,32);
+	lum_im = pixCreate(width,height,32);
+	feat_im = pixCreate(width,height,32);
+	int i,j;
+        char write_path[500];
```
Comment (Abhinav95) [inline src/lib_ccx/hardsubx_russian.c]: Indent

File: src/lib_ccx/params.c [hunk 7]
```diff
@@ -1198,6 +1198,11 @@ int parse_parameters (struct ccx_s_options *opt, int argc, char *argv[])
 			opt->hardsubx = 1;
 			continue;
 		}
+		if (strcmp(argv[i], "-letterwise_russian")==0)
+		{
+			opt->letter_russian = 1;
+			continue;
```
Comment (Abhinav95) [inline src/lib_ccx/params.c]: Put all the things related to Russian OCR together in params.c with clear commenting about what each is. Also add the usage instructions to the help screen text.

File: src/lib_ccx/hardsubx_imgops.c [hunk 8]
```diff
@@ -105,4 +105,50 @@ void rgb_to_lab(float R, float G, float B,float *L, float *a, float *b)
 	*b = YELLOW;
 }
 
+void rgb2lab(float R, float G, float B,float *L, float *a, float *b){
```
Comment (Abhinav95) [inline src/lib_ccx/hardsubx_imgops.c]: Just use the old rgb_to_lab function, why create a duplicate?

File: src/lib_ccx/hardsubx.c [hunk 9]
```diff
@@ -255,6 +260,17 @@ struct lib_hardsubx_ctx* _init_hardsubx(struct ccx_s_options *options)
 	ctx->write_format = options->write_format;
 	ctx->subs_delay = options->subs_delay;
 	ctx->cc_to_stdout = options->cc_to_stdout;
+	ctx->russian = options->russian;
```
Comment (Abhinav95) [inline src/lib_ccx/hardsubx.c]: Probably better to create a separate context for all Russian related variables only (only if Russian mode is activated).
ctx->russian->frame_skip, ctx->russian->start_frame etc instead of just having the options in the main ctx.

File: src/lib_ccx/hardsubx.c [hunk 10]
```diff
@@ -102,11 +102,16 @@ int hardsubx_process_data(struct lib_hardsubx_ctx *ctx)
 	
 	mprint("Beginning burned-in subtitle detection...\n");
 
-	if(ctx->tickertext)
+	if(ctx->tickertext && ctx->russian)
+		hardsubx_process_frames_tickertext_russian(ctx, enc_ctx);
+	else if(ctx->tickertext)
 		hardsubx_process_frames_tickertext(ctx, enc_ctx);
+	else if(ctx->russian)
```
Comment (Abhinav95) [inline src/lib_ccx/hardsubx.c]: ctx->russian without tickertext being a separate function makes no sense. We can simply set the ocrlang parameter to rus.

File: src/lib_ccx/hardsubx_russian.c [hunk 11]
```diff
@@ -0,0 +1,995 @@
...
+//TODO: Correct FFMpeg integration
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavutil/imgutils.h>
+#include <libswscale/swscale.h>
+#include "allheaders.h"
+#include "hardsubx.h"
+#include "capi.h"
+
+//Global Flags
```
Comment (Abhinav95) [inline src/lib_ccx/hardsubx_russian.c]: Instead of being globals, can these not be made part of a separate Russian context? Helps to keep track too.

File: src/lib_ccx/hardsubx_decoder.c (hunk 4 above)
Comment (BurhanUlTayyab) [reply src/lib_ccx/hardsubx_decoder.c]: Yep, This is a mistake, I did it when I was testing, shouldn't have changed that in hardsubx_decoder.c

File: src/lib_ccx/hardsubx_russian.c (hunk 5 above)
Comment (BurhanUlTayyab) [reply src/lib_ccx/hardsubx_russian.c]: Okay, Noted :)

File: src/lib_ccx/hardsubx_russian.c (hunk 6 above)
Comment (BurhanUlTayyab) [reply src/lib_ccx/hardsubx_russian.c]: Noted

File: src/lib_ccx/params.c (hunk 7 above)
Comment (BurhanUlTayyab) [reply src/lib_ccx/params.c]: Noted

File: src/lib_ccx/hardsubx_imgops.c (hunk 8 above)
Comment (BurhanUlTayyab) [reply src/lib_ccx/hardsubx_imgops.c]: The old rgb_to_lab function doesn't cater the Russian News Tickers, creates a very dull image, I rewrote the function with little changes, change in threshold basically to get a great thresholded text

File: src/lib_ccx/hardsubx.c (hunk 9 above)
Comment (BurhanUlTayyab) [reply src/lib_ccx/hardsubx.c]: Noted

File: src/lib_ccx/hardsubx.c (hunk 10 above)
Comment (BurhanUlTayyab) [reply src/lib_ccx/hardsubx.c]: It was to run just simple basic frameskip if hardsubx is activated, but noted :)

File: src/lib_ccx/hardsubx_russian.c (hunk 11 above)
Comment (BurhanUlTayyab) [reply src/lib_ccx/hardsubx_russian.c]: I've tried that, but some values just need to be initialized and changed in more than one function and there are a lot of loops that changes the value. I first tried to do that but it was not possible

File: src/lib_ccx/hardsubx_imgops.c (hunk 8 above)
Comment (Abhinav95) [reply src/lib_ccx/hardsubx_imgops.c]: Not good practice at all. You can do the same inside the russian function without creating a duplicate call which future contributors may confuse as standard.

Comment (BurhanUlTayyab) [reply src/lib_ccx/hardsubx_imgops.c]: Noted :)

**Comment by BurhanUlTayyab** [comment]
Hello Abhinav,

I have done the changes as you've guided me to, kindly take a look at it

Thanks

**Comment by Abhinav95** [comment]
Please make the indentation uniform at all places.

I'll do a second thorough review after Saurabh's functional code and yours have been merged. But please go ahead with your work product submission, demo and usage instructions included.

**Comment by BurhanUlTayyab** [comment]
Yeah Sure :) Thanks :)

**Comment by cfsmp3** [comment]
@Abhinav95 should we just close this PR? Looks like @BurhanUlTayyab is not going to finish...

**Comment by cfsmp3** [comment]
@Abhinav95 @BurhanUlTayyab Can I merge this? The other guy clearly doesn't care (a GSoC hit-and-run) so either we discard this or we do something about it so it doesn't go to waste.

**Comment by BurhanUlTayyab** [comment]
@cfsmp3 @Abhinav95 This PR works with high accuracy even without text detector, and it's already been integrated into Red Hen Lab Pipelines. So should I start working on the detector myself? Because the other guy is gone
